File: core/mobile/test/engine/fifo_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:nepse_buy/domain/engine/fifo.dart';

void main() {
  group('fifoMatch', () {
    test('consumes oldest lots first and computes per-match profit', () {
      final lots = [
        BuyLot(
          tradeId: 'l1',
          executedAt: DateTime.utc(2024, 1, 1),
          netPricePerShare: 100,
          remainingShares: 10,
        ),
        BuyLot(
          tradeId: 'l2',
          executedAt: DateTime.utc(2024, 6, 1),
          netPricePerShare: 110,
          remainingShares: 10,
        ),
      ];
      final matches = fifoMatch(
        lots: lots,
        sellShares: 15,
        sellGrossPricePerShare: 120,
        sellExecutedAt: DateTime.utc(2025, 1, 5),
        longTermDaysThreshold: 365,
      );
      expect(matches.length, 2);
      expect(matches[0].buyTradeId, 'l1');
      expect(matches[0].matchedShares, 10);
      expect(matches[0].realizedProfit, 200); // 10 * (120-100)
      expect(matches[0].isLongTerm, isTrue); // 370 days
      expect(matches[1].buyTradeId, 'l2');
      expect(matches[1].matchedShares, 5);
      expect(matches[1].realizedProfit, 50); // 5 * (120-110)
      expect(matches[1].isLongTerm, isFalse); // 218 days
      expect(lots[0].remainingShares, 0);
      expect(lots[1].remainingShares, 5);
    });

    test('throws when inventory is short', () {
      final lots = [
        BuyLot(
          tradeId: 'l1',
          executedAt: DateTime.utc(2024, 1, 1),
          netPricePerShare: 100,
          remainingShares: 3,
        ),
      ];
      expect(
        () => fifoMatch(
          lots: lots,
          sellShares: 5,
          sellGrossPricePerShare: 120,
          sellExecutedAt: DateTime.utc(2024, 6, 1),
          longTermDaysThreshold: 365,
        ),
        throwsA(isA<InsufficientInventoryException>()),
      );
    });
  });
}

File: core/backend/src/trades/__tests__/fifo.test.ts
import { describe, expect, it } from "vitest";
import { fifoMatch } from "../fifo.js";
import type { BuyLot } from "../types.js";

function lot(tradeId: string, executedAt: string, netPricePerShare: number, remainingShares: number): BuyLot {
  return { tradeId, executedAt: new Date(executedAt), netPricePerShare, remainingShares };
}

describe("fifoMatch", () => {
  it("consumes the oldest lot first when it fully covers", () => {
    const lots: BuyLot[] = [
      lot("b1", "2026-01-01", 1000, 100),
      lot("b2", "2026-03-01", 1100, 100),
    ];
    const matches = fifoMatch(lots, 50, 1300, new Date("2026-05-01"), 365);
    expect(matches).toHaveLength(1);
    expect(matches[0].buyTradeId).toBe("b1");
    expect(matches[0].matchedShares).toBe(50);
    expect(matches[0].realizedProfit).toBe(50 * 300);
    expect(lots[0].remainingShares).toBe(50);
    expect(lots[1].remainingShares).toBe(100);
  });

  it("spills across lots when one isn't enough", () => {
    const lots: BuyLot[] = [
      lot("b1", "2026-01-01", 1000, 30),
      lot("b2", "2026-03-01", 1100, 100),
    ];
    const matches = fifoMatch(lots, 50, 1300, new Date("2026-05-01"), 365);
    expect(matches).toHaveLength(2);
    expect(matches[0]).toMatchObject({ buyTradeId: "b1", matchedShares: 30 });
    expect(matches[1]).toMatchObject({ buyTradeId: "b2", matchedShares: 20 });
    expect(lots[0].remainingShares).toBe(0);
    expect(lots[1].remainingShares).toBe(80);
  });

  it("classifies long-term vs short-term per lot", () => {
    const lots: BuyLot[] = [
      lot("b1", "2024-01-01", 1000, 30), // ~500+ days back
      lot("b2", "2026-04-01", 1200, 100), // ~30 days back
    ];
    const matches = fifoMatch(lots, 50, 1300, new Date("2026-05-01"), 365);
    expect(matches[0].isLongTerm).toBe(true);
    expect(matches[1].isLongTerm).toBe(false);
  });

  it("skips lots already fully consumed", () => {
    const lots: BuyLot[] = [
      lot("b1", "2026-01-01", 1000, 0),
      lot("b2", "2026-03-01", 1100, 100),
    ];
    const matches = fifoMatch(lots, 30, 1300, new Date("2026-05-01"), 365);
    expect(matches).toHaveLength(1);
    expect(matches[0].buyTradeId).toBe("b2");
  });

  it("throws when buy inventory is insufficient", () => {
    const lots: BuyLot[] = [lot("b1", "2026-01-01", 1000, 10)];
    expect(() => fifoMatch(lots, 50, 1300, new Date("2026-05-01"), 365))
      .toThrow(/Insufficient buy inventory.*40/);
  });
});
